### PR TITLE
add TypeLimitedBlockstore

### PR DESCRIPTION
**Pull request by kevodwyer**

### CODE REVIEW

File: src/main/java/org/peergos/APIService.java [hunk 1]
```diff
@@ -16,7 +17,7 @@ public class APIService {
     private final BlockService remoteBlocks;
 
     public APIService(Blockstore store, BlockService remoteBlocks) {
-        this.store = store;
+        this.store = new TypeLimitedBlockstore(store, Set.of(Cid.Codec.Raw, Cid.Codec.DagCbor));
```
Comment (ianopolous) [inline src/main/java/org/peergos/APIService.java:20]: we shouldn't force everyone to use it.

Comment (ianopolous) [reply src/main/java/org/peergos/APIService.java:20]: maybe put it in the config, and default to not limiting it?

File: src/main/java/org/peergos/blockstore/TypeLimitedBlockstore.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        LOG.log(Level.WARNING, "Failed attempt to call bloomAdd with codec: " + cid.codec);
+        return CompletableFuture.completedFuture(false);
+    }
+
+    @Override
+    public CompletableFuture<Boolean> has(Cid cid) {
+        if (allowedCodecs.contains(cid.codec)) {
+            return blocks.has(cid);
+        }
+        LOG.log(Level.WARNING, "Failed attempt to call has with codec: " + cid.codec);
```
Comment (ianopolous) [inline src/main/java/org/peergos/blockstore/TypeLimitedBlockstore.java:39]: This and get will be called millions of times via the bitswap. So I wouldn't log it

File: src/main/java/org/peergos/blockstore/TypeLimitedBlockstore.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        return CompletableFuture.completedFuture(Optional.empty());
+    }
+
+    @Override
+    public CompletableFuture<Cid> put(byte[] block, Cid.Codec codec) {
+        if (allowedCodecs.contains(codec)) {
+            return blocks.put(block, codec);
+        }
+        LOG.log(Level.WARNING, "Failed attempt to call put with codec: " + codec);
+        Cid cid = new Cid(1, codec, Multihash.Type.sha2_256, Hash.sha256(block));
```
Comment (ianopolous) [inline src/main/java/org/peergos/blockstore/TypeLimitedBlockstore.java:58]: return error here

File: src/main/java/org/peergos/APIService.java (hunk 1 above)
Comment (kevodwyer) [reply src/main/java/org/peergos/APIService.java]: done

File: src/main/java/org/peergos/blockstore/TypeLimitedBlockstore.java (hunk 2 above)
Comment (kevodwyer) [reply src/main/java/org/peergos/blockstore/TypeLimitedBlockstore.java]: logging removed from get and has

File: src/main/java/org/peergos/blockstore/TypeLimitedBlockstore.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    public TypeLimitedBlockstore(Blockstore blocks, Set<Cid.Codec> allowedCodecs) {
+        this.blocks = blocks;
+        this.allowedCodecs = allowedCodecs;
+    }
+
+    public CompletableFuture<Boolean> bloomAdd(Cid cid) {
+        if (allowedCodecs.contains(cid.codec)) {
+            return blocks.bloomAdd(cid);
+        }
+        LOG.log(Level.WARNING, "Failed attempt to call bloomAdd with codec: " + cid.codec);
```
Comment (ianopolous) [inline src/main/java/org/peergos/blockstore/TypeLimitedBlockstore.java:28]: Adding a banned codec should be an error

File: src/main/java/org/peergos/net/APIHandler.java [hunk 5]
```diff
@@ -128,7 +128,8 @@ public void handleCallToAPI(HttpExchange httpExchange) {
                     if (block.length >  1024 * 1024 * 2) { //todo what should the limit be?
                         throw new APIException("Block too large");
                     }
-                    Cid cid = service.putBlock(block, formatOpt.get());
+                    Cid.Codec codec = formatOpt.get().equals("raw") ? Cid.Codec.Raw : Cid.Codec.DagCbor;
```
Comment (ianopolous) [inline src/main/java/org/peergos/net/APIHandler.java:131]: Api handler shoudn't limit the codecs

File: src/main/java/org/peergos/Server.java [hunk 6]
```diff
@@ -93,7 +93,9 @@ public Server() throws Exception {
         int handlerThreads = 50;
         LOG.info("Starting RPC API server at: localhost:" + localAPIAddress.getPort());
         HttpServer apiServer = HttpServer.create(localAPIAddress, maxConnectionQueue);
-        APIService service = new APIService(blockStore, new BitswapBlockService(node, builder.getBitswap().get()));
+        Blockstore apiServiceBlockstore = config.datastore.allowedCodecs.codecs.isEmpty() ?
```
[the remark this answers is not in the document]
Comment (ianopolous) [reply src/main/java/org/peergos/Server.java:96]: I would move all the blockstore construction to a separate method

File: src/main/java/org/peergos/net/APIHandler.java (hunk 5 above)
Comment (kevodwyer) [reply src/main/java/org/peergos/net/APIHandler.java]: fixed

File: src/main/java/org/peergos/APIService.java [hunk 7]
```diff
@@ -42,16 +43,17 @@ public List<HashedBlock> getBlocks(List<Want> wants, Set<PeerId> peers, boolean
...
-    
-    public Cid putBlock(byte[] block, String format) {
-        Cid.Codec codec = null;
-        if (format.equals("raw")) {
-            codec = Cid.Codec.Raw;
-        } else if (format.equals("cbor")) {
-            codec = Cid.Codec.DagCbor;
+
+    public boolean accepts(Cid.Codec codec) {
+        if (store instanceof TypeLimitedBlockstore) {
```
Comment (ianopolous) [inline src/main/java/org/peergos/APIService.java:48]: This class shouldn't depend on a specific blockstore impl. Just parse the codec from the strings.

**Review by ianopolous — APPROVED** [review]

**Pull request merged**